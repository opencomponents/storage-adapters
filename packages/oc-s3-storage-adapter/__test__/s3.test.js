'use strict';
const s3 = require('../');
const AWS = require('aws-sdk');

//Mock AWS functions
AWS.S3.prototype.getObject = (val, cb) => {
  cb(null, {
    Body: '{"data": "words"}'
  });
};

AWS.S3.prototype.listObjects = (val, cb) => {
  cb(null, {
    CommonPrefixes: [
      {
        Prefix: '/testPrefix/'
      },
      {
        Prefix: '/testPrefix2/'
      }
    ]
  });
};

//Mock Date functions
const DATE_TO_USE = new Date('2017');
const _Date = Date;
global.Date = jest.fn(() => DATE_TO_USE);
global.Date.UTC = _Date.UTC;
global.Date.parse = _Date.parse;
global.Date.now = _Date.now;

//Tests
test('should expose the correct methods', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    key: 'test-key',
    secret: 'test-secret'
  };
  expect(new s3(options)).toMatchSnapshot();
});

test('validate valid conf', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    key: 'test-key',
    secret: 'test-secret'
  };
  const client = new s3(options);
  expect(client.isValid()).toBe(true);
});

test('validate missing bucket conf', () => {
  const options = {
    region: 'region-test',
    key: 'test-key',
    secret: 'test-secret'
  };
  const client = new s3(options);
  expect(client.isValid()).toBe(false);
});

test('validate missing region conf', () => {
  const options = {
    bucket: 'test',
    key: 'test-key',
    secret: 'test-secret'
  };
  const client = new s3(options);
  expect(client.isValid()).toBe(false);
});

test('validate missing key conf', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    secret: 'test-secret'
  };
  const client = new s3(options);
  expect(client.isValid()).toBe(false);
});

test('validate missing secret conf', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    key: 'test-key'
  };
  const client = new s3(options);
  expect(client.isValid()).toBe(false);
});

test('validate missing key/secret conf', () => {
  const missingOptions = {
    bucket: 'test',
    region: 'region-test'
  };
  const client = new s3(missingOptions);
  expect(client.isValid()).toBe(true);
});

//Functions utilizing AWS

test('test getFile ', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    key: 'test-key',
    secret: 'test-secret'
  };
  const client = new s3(options);
  const cb = (err, data) => {
    expect(data).toMatchSnapshot();
  };
  client.getFile('path/test.json', false, cb);
});

test('test getJSon ', () => {
  const options = {
    bucket: 'test',
    region: 'region-test',
    key: 'test-key',
    secret: 'test-secret',
    agentProxy: 'agentProxy'
  };
  const client = new s3(options);
  const cb = (err, data) => {
    expect(data).toMatchSnapshot();
  };
  client.getJson('path/test.json', false, cb);
});

test('test listObjects ', () => {
  const client = new s3({ bucket: 'my-bucket' });
  const cb = (err, data) => {
    expect(data).toMatchSnapshot();
  };
  client.listSubDirectories('path/', cb);
});

test('test getUrl ', () => {
  const client = new s3({ path: '/' });
  expect(client.getUrl('test', '1.0.0', 'test.js')).toMatchSnapshot();
});

test('test private putFileContent ', () => {
  const client = new s3({ bucket: 'my-bucket' });
  const cb = data => {
    expect(data).toMatchSnapshot();
  };

  AWS.S3.prototype.upload = data => ({ send: fn => fn(cb(data)) });
  client.putFileContent('words', 'filename.js', true, cb);
});

test('test public putFileContent ', () => {
  const client = new s3({ bucket: 'my-bucket' });
  const cb = data => {
    expect(data).toMatchSnapshot();
  };

  AWS.S3.prototype.upload = data => ({ send: fn => fn(cb(data)) });
  client.putFileContent('words', 'filename.gz', false, cb);
});

test('test putFile ', () => {
  const client = new s3({ bucket: 'my-bucket' });
  const cb = data => {
    expect(data).toMatchSnapshot();
  };

  AWS.S3.prototype.upload = data => ({ send: fn => fn(cb(data)) });
  client.putFile('package.json', 'filename.js', false, cb);
});
