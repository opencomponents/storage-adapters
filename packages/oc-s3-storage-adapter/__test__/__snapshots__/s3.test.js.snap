// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expose the correct methods 1`] = `
Object {
  "adapterType": "s3",
  "getFile": [Function],
  "getJson": [Function],
  "getUrl": [Function],
  "isValid": [Function],
  "listSubDirectories": [Function],
  "maxConcurrentRequests": 20,
  "putDir": [Function],
  "putFile": [Function],
  "putFileContent": [Function],
}
`;

exports[`test getFile  1`] = `"{\\"data\\": \\"words\\"}"`;

exports[`test getJSon  1`] = `
Object {
  "data": "words",
}
`;

exports[`test getUrl  1`] = `"/test/1.0.0/test.js"`;

exports[`test listObjects  1`] = `
Array [
  "Prefix",
  "Prefix2",
]
`;

exports[`test private putFileContent  1`] = `
Object {
  "ACL": "authenticated-read",
  "Body": "words",
  "Bucket": "my-bucket",
  "ContentType": "application/javascript",
  "Expires": 2018-01-01T00:00:00.000Z,
  "Key": "filename.js",
  "ServerSideEncryption": "AES256",
}
`;

exports[`test private putFileContent  2`] = `undefined`;

exports[`test public putFileContent  1`] = `
Object {
  "ACL": "public-read",
  "Body": "words",
  "Bucket": "my-bucket",
  "ContentEncoding": "gzip",
  "Expires": 2019-01-01T00:00:00.000Z,
  "Key": "filename.gz",
  "ServerSideEncryption": "AES256",
}
`;

exports[`test public putFileContent  2`] = `undefined`;

exports[`test putFile  1`] = `
Object {
  "ACL": "public-read",
  "Body": ReadStream {
    "_events": Object {
      "end": [Function],
    },
    "_eventsCount": 1,
    "_maxListeners": undefined,
    "_readableState": ReadableState {
      "awaitDrain": 0,
      "buffer": BufferList {
        "head": null,
        "length": 0,
        "tail": null,
      },
      "decoder": null,
      "defaultEncoding": "utf8",
      "destroyed": false,
      "emittedReadable": false,
      "encoding": null,
      "endEmitted": false,
      "ended": false,
      "flowing": null,
      "highWaterMark": 65536,
      "length": 0,
      "needReadable": false,
      "objectMode": false,
      "pipes": null,
      "pipesCount": 0,
      "readableListening": false,
      "reading": false,
      "readingMore": false,
      "resumeScheduled": false,
      "sync": true,
    },
    "autoClose": true,
    "bytesRead": 0,
    "domain": null,
    "end": undefined,
    "fd": null,
    "flags": "r",
    "mode": 438,
    "path": "package.json",
    "pos": undefined,
    "readable": true,
    "start": undefined,
  },
  "Bucket": "my-bucket",
  "ContentType": "application/javascript",
  "Expires": 2020-01-01T00:00:00.000Z,
  "Key": "filename.js",
  "ServerSideEncryption": "AES256",
}
`;

exports[`test putFile  2`] = `undefined`;
